docs/contributing/index.rst

```diff
@@ -1,9 +1,114 @@
 Contributing
 ============
 
-Contributing to RestrictedPython 4+
+Contributing to RestrictedPython
 
 Todos
 -----
 
 .. todolist::
+
+Understanding How RestrictedPython works internally
+---------------------------------------------------
+
+RestrictedPython is a classic approach of compiler construction to create a limited subset of an existing programming language.
+
+Defining a programming language requires a regular grammar (`Chomsky 3`_ / `EBNF`_) definition.
+This grammar will be implemented in an abstract syntax tree (AST), which will be passed on to a code generator to produce a machine-readable version.
+
+.. _`_sec_code_generation`:
+
+Code generation
++++++++++++++++
+
+As Python is a platform independent programming language, this machine readable version is a byte code which will be translated on the fly by an interpreter into machine code.
+This machine code then gets executed on the specific CPU architecture, with the standard operating system restrictions.
+
+The byte code produced must be compatible with the execution environment that the Python interpreter is running in, so we do not generate the byte code directly from ``compile_restricted`` and the other ``compile_restricted_*`` methods manually, it may not match what the interpreter expects.
+
+Thankfully, the Python ``compile()`` function introduced the capability to compile ``ast.AST`` code into byte code in Python 2.6, so we can return the platform-independent AST and keep byte code generation delegated to the interpreter.
+
+``ast`` module (Abstract Syntax Trees)
+++++++++++++++++++++++++++++++++++++++
+
+The ``ast`` module consists of four areas:
+
+* ``AST`` (Basis of all Nodes) + all node class implementations
+* ``NodeVisitor`` and ``NodeTransformer`` (tool to consume and modify the AST)
+* Helper methods
+
+  * ``parse``
+  * ``walk``
+  * ``dump``
+
+* Constants
+
+  * ``PyCF_ONLY_AST``
+
+
+``NodeVisitor`` & ``NodeTransformer``
++++++++++++++++++++++++++++++++++++++
+
+A ``NodeVisitor`` is a class of a node / AST consumer, it reads the data by stepping through the tree without modifying it.
+In contrast, a ``NodeTransformer`` (which inherits from a ``NodeVisitor``) is allowed to modify the tree and nodes.
+
+Modifying the AST
++++++++++++++++++
+
+
+Technical Backgrounds - Links to External Documentation
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+* Concept of Immutable Types and Python Example (https://en.wikipedia.org/wiki/Immutable_object#Python)
+* Python 3 Standard Library Documentation on AST module ``ast`` (https://docs.python.org/3/library/ast.html)
+
+  * AST Grammar of Python
+
+    * `Python 3.8 AST`_
+    * `Python 3.7 AST`_
+    * `Python 3.6 AST`_
+    * `Python 3.5 AST`_
+    * `Python 3.4 AST`_ (obsolete)
+    * `Python 3.3 AST`_ (obsolete)
+    * `Python 3.2 AST`_ (obsolete)
+    * `Python 3.1 AST`_ (obsolete)
+    * `Python 3.0 AST`_ (obsolete)
+    * `Python 2.7 AST`_
+    * `Python 2.6 AST`_ (obsolete)
```

Comment:
I am not sure if it helps to list so many obsolete Python grammars.

Reply:
with the current supported / used versions that helps, we can later remove some, which we don't need anymore, when we drop finally support for Python 2

Reply:
okay